### PR TITLE
kubeadm: skip Swap check if failSwapOn is false

### LINKED ISSUE #1254: failSwapOn in kubeadm-config file does not work
/bug

[root@host-10-10-10-33 ~]# cat /tmp/kubeadm-config.yml
~~~
apiVersion: kubeadm.k8s.io/v1alpha3
kind: ClusterConfiguration
kubernetesVersion: v1.12.2
apiServerCertSANs:
\- "10.10.10.33"
controlPlaneEndpoint: "10.10.10.33:6443"
etcd:
  local:
    extraArgs:
      listen-client-urls: "https://127.0.0.1:2379,https://10.10.10.33:2379"
      advertise-client-urls: "https://10.10.10.33:2379"
      listen-peer-urls: "https://10.10.10.33:2380"
      initial-advertise-peer-urls: "https://10.10.10.33:2380"
      initial-cluster: >-
        "
        host-10-10-10-33.openstacklocal=https://10.10.10.33:2380
        "
    serverCertSANs:
      - "host-10-10-10-33.openstacklocal"
      - "10.10.10.33"
    peerCertSANs:
      - "host-10-10-10-33.openstacklocal"
      - "10.10.10.33"

networking:
    \# This CIDR is a Calico default. Substitute or remove for your CNI provider.
    podSubnet: "192.168.0.0/16"
imageRepository: apphub.zte.com.cn/itp/kubernetes
unifiedControlPlaneImage: ""
---
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
failSwapOn: false
~~~

[root@host-10-10-10-33 ~]# kubeadm init --config=/tmp/kubeadm-config.yml

> [init] using Kubernetes version: v1.12.2
[preflight] running pre-flight checks
[preflight] Some fatal errors occurred:
        [ERROR Swap]: running with swap on is not supported. Please disable swap
[preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors=...`

### PR DESCRIPTION
**Pull request by bart0sh**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

It makes sense to skip Swap preflight check If user sets
failSwapOn: false in the config when running kubeadm init.

**Which issue(s) this PR fixes**:

Fixes: kubernetes/kubeadm#1254

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: kubeadm init skips Swap check if failSwapOn is set to false in user config
```

### CODE REVIEW

File: cmd/kubeadm/app/preflight/checks.go [hunk 1]
```diff
@@ -940,6 +945,7 @@ func RunJoinNodeChecks(execer utilsexec.Interface, cfg *kubeadmapi.JoinConfigura
 		DirAvailableCheck{Path: filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ManifestsSubDirName)},
 		FileAvailableCheck{Path: filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.KubeletKubeConfigFileName)},
 		FileAvailableCheck{Path: filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.KubeletBootstrapKubeConfigFileName)},
+		SwapCheck{},
```
Comment (neolit123) [inline cmd/kubeadm/app/preflight/checks.go]: this as to be wrapped for OS == linux only.

unfortunately we always run it on Join(), because we ignore the KubeletConfiguration on "join".
but this is out of scope for this patch.

File: cmd/kubeadm/app/preflight/checks.go [hunk 2]
```diff
@@ -886,6 +886,11 @@ func RunInitMasterChecks(execer utilsexec.Interface, cfg *kubeadmapi.InitConfigu
 		HTTPProxyCIDRCheck{Proto: "https", CIDR: cfg.Networking.ServiceSubnet},
 		HTTPProxyCIDRCheck{Proto: "https", CIDR: cfg.Networking.PodSubnet},
 	}
+
+	if cfg.ComponentConfigs.Kubelet == nil || cfg.ComponentConfigs.Kubelet.FailSwapOn {
+		checks = append(checks, SwapCheck{})
```
Comment (neolit123) [inline cmd/kubeadm/app/preflight/checks.go]: it has to happen for linux only.

something else we can do for both init and join is to check if the `fail-swap-on` flag exists in the KubeletExtraArgs in NodeRegistrationOptions for both Init and JoinConfiguration:
https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1#NodeRegistrationOptions

such a flag should override the value of cfg.ComponentConfigs.Kubelet.FailSwapOn.
currently cfg.ComponentConfigs.Kubelet is only available for init.

**Review by neolit123 — APPROVED** [review]
@bart0sh thanks.
added a couple of comments.

/priority backlog
/assign @fabriziopandini 
this might not make it in 1.13 as its not a critical fix.

File: cmd/kubeadm/app/preflight/checks.go (hunk 2 above)
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks.go]: > it has to happen for linux only.

True. will do.

> something else we can do

Does this make sense to do it in this PR or separately?

Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks.go]: i think it should be in the same PR.

Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks.go]: > something else we can do for both init and join is to check if the fail-swap-on flag exists in the KubeletExtraArgs 

I'm not sure we should do it as the flag seems to be deprecated:
```
--fail-swap-on  Makes the Kubelet fail to start if swap is enabled on the node. 
(default true) (DEPRECATED: This parameter should be set via the config
file specified by the Kubelet's --config flag.
See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
```

File: cmd/kubeadm/app/preflight/checks.go (hunk 1 above)
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks.go]: added OS == linux checks in both places. Thank you for pointing this out!

File: cmd/kubeadm/app/preflight/checks.go (hunk 2 above)
Comment (neolit123) [reply cmd/kubeadm/app/preflight/checks.go]: as far as we know some of these flags could be a around for "a while" and we depend on them for the kubelet env file.

even if it  leaves us without an option to check if the Join node wants to skip, i think it's OK to to not do it for this reason.

instead i think the joining node should eventually have a custom KubeletConfiguration but that's outside of scope for this PR.

Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks.go]: > even if it leaves us without an option to check if the Join node wants to skip, i think it's OK to to not do it for this reason

sorry for misunderstanding, so should we do it or not?

If not - please lgtm

**Comment by timothysc** [comment]
/hold 

Let's talk about this one.

**Comment by bart0sh** [comment]
@timothysc 

> Let's talk about this one.

Let's talk. Where and when?

**Comment by fabriziopandini** [comment]
@bart0sh
IMO best option for discussing this point is kubeadm office hours/next week
This week almost all the people are at kubeconf, and most probably weely meeting will be suspended due to lack of quorum (we will sink on slack).

**Comment by bart0sh** [comment]
@fabriziopandini thanks. I'll try to be there.

**Comment by chuckha** [comment]
I don't see this in our notes, was this ever discussed at a meeting?

From the perspective of kubeadm having production ready defaults, it feels a bit odd to ignore a check that is important for production clusters. Since this flag would likely only be used in test-situations anyway, running kubeadm with `--ignore-preflight-errors=Swap` would be enough for people to get past this in their test setting and make it even more challenging to accidentally end up with a cluster that has non-production flags enabled.

**Comment by neolit123** [comment]
> I don't see this in our notes, was this ever discussed at a meeting?

@bart0sh can you please add this to the agenda for next week - 30th Jan?

**Comment by bart0sh** [comment]
@neolit123 done

**Comment by bart0sh** [comment]
Closing as it's not going to be approved.
